File: backend/process_utils.py
"""Windows process helpers used by the launch pipeline - tasklist checks,
starting an app (with UAC elevation fallback), killing a PID, and building
the right Steam/MS Store launch command for MSFS.
"""
import os
import subprocess

from . import win_native

_MSFS_LAUNCH_COMMANDS = {
    ("MSFS 2020", "Steam"): "start steam://rungameid/1250410",
    ("MSFS 2020", "MS Store"): "start shell:AppsFolder\\Microsoft.FlightSimulator_8wekyb3d8bbwe!App",
    ("MSFS 2024", "Steam"): "start steam://rungameid/2537590",
    ("MSFS 2024", "MS Store"): "start shell:AppsFolder\\Microsoft.FlightSimulator2024_8wekyb3d8bbwe!App",
}
_DEFAULT_MSFS_LAUNCH_COMMAND = "start steam://rungameid/2537590"


def tasklist_lower():
    try:
        return os.popen("tasklist").read().lower()
    except Exception:
        return ""


def is_msfs_running(tasks):
    return "flightsimulator.exe" in tasks or "flightsimulator2024.exe" in tasks or "flightsimulator" in tasks


def is_exe_running(path, tasks):
    return os.path.basename(path).lower() in tasks


def run_app(path, as_admin=False):
    """Returns the launched PID, or None on failure - mirrors the old run_app/_run_as_admin."""
    if not os.path.exists(path):
        return None
    try:
        if as_admin:
            return win_native.run_as_admin(path, cwd=os.path.dirname(path) or None)
        proc = subprocess.Popen([path], cwd=os.path.dirname(path) or None)
        return proc.pid
    except OSError as e:
        if getattr(e, "winerror", None) == 740:
            # ERROR_ELEVATION_REQUIRED - the exe demands admin via its own manifest
            return win_native.run_as_admin(path, cwd=os.path.dirname(path) or None)
        return None
    except Exception:
        return None


def terminate_pid(pid):
    try:
        os.system(f"taskkill /PID {pid} /F /T >nul 2>&1")
    except Exception:
        pass


def launch_msfs(sim_version, sim_platform):
    command = _MSFS_LAUNCH_COMMANDS.get((sim_version, sim_platform), _DEFAULT_MSFS_LAUNCH_COMMAND)
    os.system(command)

File: dotnet/src/FlightOpsHub.Core/ProcessUtils.cs
using System.ComponentModel;
using System.Diagnostics;

namespace FlightOpsHub.Core;

/// <summary>
/// Port of backend/process_utils.py - Windows process helpers used by the
/// launch pipeline. Simpler than the Python original in one respect: .NET's
/// ProcessStartInfo(UseShellExecute=true, Verb="runas") triggers the UAC
/// elevation prompt natively, so no win32/ctypes ShellExecuteEx P/Invoke
/// (win_native.py's run_as_admin) is needed here at all.
/// </summary>
public static class ProcessUtils
{
    private static readonly Dictionary<(string SimVersion, string SimPlatform), string> MsfsLaunchCommands = new()
    {
        [("MSFS 2020", "Steam")] = "steam://rungameid/1250410",
        [("MSFS 2020", "MS Store")] = @"shell:AppsFolder\Microsoft.FlightSimulator_8wekyb3d8bbwe!App",
        [("MSFS 2024", "Steam")] = "steam://rungameid/2537590",
        [("MSFS 2024", "MS Store")] = @"shell:AppsFolder\Microsoft.FlightSimulator2024_8wekyb3d8bbwe!App",
    };
    private const string DefaultMsfsLaunchCommand = "steam://rungameid/2537590";

    public static string TasklistLower()
    {
        try
        {
            using var proc = Process.Start(new ProcessStartInfo("tasklist")
            {
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            });
            var output = proc!.StandardOutput.ReadToEnd();
            proc.WaitForExit();
            return output.ToLowerInvariant();
        }
        catch
        {
            return "";
        }
    }

    public static bool IsMsfsRunning(string tasksLower) =>
        tasksLower.Contains("flightsimulator.exe") || tasksLower.Contains("flightsimulator2024.exe") || tasksLower.Contains("flightsimulator");

    public static bool IsExeRunning(string path, string tasksLower) =>
        tasksLower.Contains(Path.GetFileName(path).ToLowerInvariant());

    /// <summary>Returns the launched PID, or null on failure.</summary>
    public static int? RunApp(string path, bool asAdmin = false)
    {
        if (!File.Exists(path)) return null;
        try
        {
            if (asAdmin) return RunAsAdmin(path);
            using var proc = Process.Start(new ProcessStartInfo(path)
            {
                WorkingDirectory = Path.GetDirectoryName(path) is { Length: > 0 } dir ? dir : null,
                UseShellExecute = false,
            });
            return proc?.Id;
        }
        catch (Win32Exception ex) when (ex.NativeErrorCode == 740) // ERROR_ELEVATION_REQUIRED - the exe demands admin via its own manifest
        {
            return RunAsAdmin(path);
        }
        catch
        {
            return null;
        }
    }

    private static int? RunAsAdmin(string path)
    {
        try
        {
            using var proc = Process.Start(new ProcessStartInfo(path)
            {
                WorkingDirectory = Path.GetDirectoryName(path) is { Length: > 0 } dir ? dir : null,
                UseShellExecute = true,
                Verb = "runas",
            });
            return proc?.Id;
        }
        catch
        {
            // User declined the UAC prompt, or the launch failed outright.
            return null;
        }
    }

    public static void TerminatePid(int pid)
    {
        try
        {
            var psi = new ProcessStartInfo("taskkill")
            {
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
            };
            psi.ArgumentList.Add("/PID");
            psi.ArgumentList.Add(pid.ToString());
            psi.ArgumentList.Add("/F");
            psi.ArgumentList.Add("/T");
            using var proc = Process.Start(psi);
            proc?.WaitForExit();
        }
        catch
        {
            // Best-effort, matches Python's bare except here.
        }
    }

    public static void LaunchMsfs(string simVersion, string simPlatform)
    {
        var command = MsfsLaunchCommands.GetValueOrDefault((simVersion, simPlatform), DefaultMsfsLaunchCommand);
        try
        {
            Process.Start(new ProcessStartInfo(command) { UseShellExecute = true });
        }
        catch
        {
            // Best-effort - matches Python's os.system() call, which also never raises.
        }
    }
}
